content/library/advanced-features/app-testing.md

```diff
@@ -0,0 +1,113 @@
+---
+title: Getting started with app testing
+slug: /library/advanced-features/app-testing
+---
+
+# Getting started with app testing
+
+Streamlit app testing enables developers to build and run headless tests that execute their app code directly, simulate user input, and inspect rendered outputs for correctness.
+
+The provided class, AppTest, simulates a running app and provides methods to set up, manipulate, and inspect the app contents via API instead of a browser UI. It can be used to write automated tests of an app in various scenarios. These can then be run using a tool like [pytest](https://docs.pytest.org/). A typical pattern is to build a suite of tests for an app to ensure consistent functionality as the app evolves. The tests run locally and/or in a CI environment like Github Actions.
+
+## Using pytest
+
+We'll use `pytest` for our examples since it is one of the most Python test frameworks. To get started, be sure to install `pytest` into your environment:
+
+```bash
+pip install pytest
+```
+
+`pytest` uses a naming convention for files to conveniently execute tests. When you write tests, name your test scripts of the form `test_<name>.py` or `<name>_test.py`. For example, `test_myapp.py` or `myapp_test.py`. When calling `pytest` in a directory, all files named accordingly will be executed to produce test results. You can place test files anywhere in your repository, but it is common to collect tests into a designated `tests/` folder. For other ways to structure and execute tests, check out [How to invoke pytest](https://docs.pytest.org/how-to/usage.html) in the `pytest` docs.
+
+Consider the following example:
+
+```none
+myproject/
+├── app.py
+└── tests/
+    └── test_app.py
+```
+
+```python
+"""app.py"""
+import streamlit as st
+
+# Initialize st.session_state.beans
+beans = st.session_state.get("beans", 0)
+
+st.title("Bean counter :paw_prints:")
+
+addend = st.number_input("Beans to add", 0, 10)
+if st.button("Add"):
+    st.session_state.beans += addend
+st.write(f"Toe beans counted: {st.session_state.beans}")
+```
+
```

Comment:
New heading for "Write a simple test" somewhere here? Don't think we want to put everything under a "using pytest" heading?

Reply:
Headers have been updated.

Reply:
LGTM